src/memory_coloring.cpp

```diff
@@ -0,0 +1,381 @@
+{
+    return std::max(x.first, y.first) < std::min(x.second, y.second);
+}
+
+struct allocation_segment
+{
+    using segment = std::pair<std::size_t, std::size_t>;
+    std::unordered_map<instruction_ref, segment> ins2segment;
+
+    const segment* add_segment(instruction_ref ins, segment s)
+    {
+        // this->remove(ins);
+        return &(ins2segment[ins] = s);
+    }
+
+    const segment* get_segment(instruction_ref ins) const
+    {
+        auto it = ins2segment.find(ins);
+        if(it == ins2segment.end())
+            return nullptr;
+        return &it->second;
+    }
+
+    // Remove segment for an instruction
+    void remove(instruction_ref ins)
+    {
+        auto it = ins2segment.find(ins);
+        if(it != ins2segment.end())
+        {
+            ins2segment.erase(it);
+        }
+    }
+
+    std::size_t max()
+    {
+        std::size_t n = 0;
+        for(auto&& pp : ins2segment)
+        {
+            auto seg = pp.second;
+            n        = std::max(n, seg.second);
+        }
+        return n;
+    }
+
+    template <class Iterator>
+    static bool overlaps(Iterator first, Iterator last, const segment& s)
+    {
+        return std::any_of(first, last, [&](auto&& t) { return is_overlap(s, t); });
+    }
+
+    static bool overlaps(const std::set<segment>& segments, const segment& s)
+    {
+        return overlaps(segments.begin(), segments.end(), s);
+    }
+
+    static auto find_gap(const std::set<segment>& segments, std::size_t n)
+    {
+        auto start = segments.begin();
+        while(true)
+        {
+            start = std::adjacent_find(start, segments.end(), [&](segment x, segment y) {
+                if(is_overlap(x, y))
+                    return false;
+                assert(y.first >= x.second);
+                auto k = y.first - x.second;
+                return (k >= n);
+            });
+            if(start == segments.end())
+                return start;
+            if(overlaps(segments.begin(), start, *start))
+            {
+                start++;
+                continue;
+            }
+            return start;
+        }
+    }
+
+    static std::size_t max_type_size(const shape& s)
+    {
+        return std::accumulate(
+            s.sub_shapes().begin(),
+            s.sub_shapes().end(),
+            s.type_size(),
+            [](auto size, const auto& sub) { return std::max(size, max_type_size(sub)); });
+    }
+
+    static std::size_t compute_alignment(instruction_ref ins)
+    {
+        auto alignment = max_type_size(ins->get_shape());
+        // A rough estimate fo the total number of elements
+        auto n = ins->get_shape().bytes() / alignment;
+        // Check for vectorized alignment
+        if(n > 4)
+        {
+            auto d = n % 4;
+            if(d == 0)
+                alignment *= 4;
+            if(d == 2)
+                alignment *= 2;
+        }
+        return alignment;
+    }
```

Comment:
Don't we want to handle the odd cases, there an assumption we add padding then?

 2, and 4 are powers of two, but are we setting the boundaries for this or is that dependent on how we read each block of memory?

Reply:
Vectorized sizes that are odd are not very common, and we currently do not vectorize these sizes.

Reply:
That makes sense.